src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -66,70 +65,183 @@ public void WriteMessage(HubMessage message, Stream output)
+                                    {
+                                        var messageType = reader.ReadAsInt32();
+
+                                        if (messageType == null)
+                                        {
+                                            throw new InvalidDataException($"Missing required property '{TypePropertyName}'.");
+                                        }
+
+                                        type = messageType.Value;
+                                    }
+                                    else if (string.Equals(memberName, InvocationIdPropertyName, StringComparison.Ordinal))
+                                    {
+                                        invocationId = reader.ReadAsString();
+                                    }
+                                    else if (string.Equals(memberName, TargetPropertyName, StringComparison.Ordinal))
+                                    {
+                                        target = reader.ReadAsString();
+                                    }
+                                    else if (string.Equals(memberName, ErrorPropertyName, StringComparison.Ordinal))
+                                    {
+                                        error = reader.ReadAsString();
+                                    }
+                                    else if (string.Equals(memberName, ResultPropertyName, StringComparison.Ordinal))
+                                    {
+                                        reader.Read();
+
+                                        if (string.IsNullOrEmpty(invocationId))
+                                        {
+                                            // If we don't have an invocation id then we need to store it as a JToken so we can parse it later
+                                            resultToken = JToken.Load(reader);
+                                        }
+                                        else
+                                        {
+                                            // If we have an invocation id already we can parse the end result
+                                            var returnType = binder.GetReturnType(invocationId);
+                                            result = PayloadSerializer.Deserialize(reader);
+                                        }
+                                    }
+                                    else if (string.Equals(memberName, ItemPropertyName, StringComparison.Ordinal))
+                                    {
+                                        reader.Read();
+
+                                        if (string.IsNullOrEmpty(invocationId))
+                                        {
+                                            // If we don't have an invocation id then we need to store it as a JToken so we can parse it later
+                                            itemToken = JToken.Load(reader);
```

Comment:
Maybe share the resultToken/result and itemToken/item variables. A message can't have both I believe.

Reply:
Yea, I just wanted to be explicit. I'll have to check if item is JToken in the methods that consume them.

Reply:
ok!

Reply:
I changed it 😄. Will see if it affects perf.

Reply:
Changing it back